cmd/portal/cmd/cmdpricing/stripe_service.go

```diff
@@ -0,0 +1,345 @@
+			var r *usage.UsageRecord
+			r, err = s.scanUsageRecord(rows)
+			if err != nil {
+				return
+			}
+			out = append(out, r)
+		}
+
+		return
+	})
+	return
+}
+
+func (s *StripeService) markStripeTimestamp(usageRecords []*usage.UsageRecord, midnight time.Time) (err error) {
+	err = s.Handle.WithTx(func() (err error) {
+		var ids []string
+		for _, record := range usageRecords {
+			ids = append(ids, record.ID)
+		}
+
+		q := s.SQLBuilder.Update(
+			s.SQLBuilder.TableName("_portal_usage_record"),
+		).
+			Set("stripe_timestamp", midnight).
+			Where("id = ANY (?)", pq.Array(ids))
+
+		result, err := s.SQLExecutor.ExecWith(q)
+		if err != nil {
+			return
+		}
+
+		rowsAffected, err := result.RowsAffected()
+		if err != nil {
+			return
+		}
+
+		if rowsAffected != int64(len(ids)) {
+			err = fmt.Errorf("failed to update usage records: %v", ids)
+			return
+		}
+
+		return
+	})
+	return
+}
+
+func (s *StripeService) uploadUsageRecordToSubscriptionItem(
+	ctx context.Context,
+	appID string,
+	subscription *stripe.Subscription,
+	si *stripe.SubscriptionItem,
+	recordName usage.RecordName,
+	midnight time.Time,
+) (err error) {
+	records, err := s.getUsageRecordsForUpload(
+		appID,
+		recordName,
+		midnight,
+	)
+	if err != nil {
+		return
+	}
+
+	// Skip when there are no records.
+	if len(records) <= 0 {
+		return
+	}
+
+	quantity := 0
+	for _, record := range records {
+		quantity += record.Count
+	}
+
+	timestamp := midnight.Unix()
+	_, err = s.ClientAPI.UsageRecords.New(&stripe.UsageRecordParams{
+		SubscriptionItem: stripe.String(si.ID),
+		Action:           stripe.String(stripe.UsageRecordActionSet),
+		Quantity:         stripe.Int64(int64(quantity)),
+		Timestamp:        &timestamp,
+	})
+	if err != nil {
+		return
+	}
+
+	err = s.markStripeTimestamp(records, midnight)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func (s *StripeService) getStripeSubscription(ctx context.Context, stripeSubscriptionID string) (*stripe.Subscription, error) {
+	params := &stripe.SubscriptionParams{
+		Params: stripe.Params{
+			Context: ctx,
+			Expand:  []*string{stripe.String("items.data.price.product")},
+		},
+	}
+	return s.ClientAPI.Subscriptions.Get(stripeSubscriptionID, params)
+}
+
+func (s *StripeService) findSubscriptionItem(subscription *stripe.Subscription, metadata map[string]string) (*stripe.SubscriptionItem, bool) {
+Loop:
+	for _, subscriptionItem := range subscription.Items.Data {
+		product := subscriptionItem.Price.Product
+
+		// Find SubscriptionItem by checking if the metadata of the product is a superset of input metadata.
+		for key, value := range metadata {
+			v := product.Metadata[key]
+			if value != v {
+				continue Loop
+			}
+		}
+
+		si := subscriptionItem
+		return si, true
+	}
+	return nil, false
+}
+
+func (s *StripeService) uploadUsage(ctx context.Context, appID string) (err error) {
+	midnight := timeutil.TruncateToDate(s.Clock.NowUTC())
+
+	stripeSubscriptionID, err := s.getStripeSubscriptionID(appID)
```

Comment:
We should handle record not found of `getStripeSubscriptionID` gracefully, since most of the apps should not have subscriptions and the command should not return error.